streams/src/main/java/org/apache/kafka/streams/TopologyConfig.java

```diff
@@ -209,19 +219,22 @@ public TopologyConfig(final String topologyName, final StreamsConfig globalAppCo
             deserializationExceptionHandlerSupplier = () -> globalAppConfigs.getConfiguredInstance(DEFAULT_DESERIALIZATION_EXCEPTION_HANDLER_CLASS_CONFIG, DeserializationExceptionHandler.class);
         }
 
-        if (isTopologyOverride(DEFAULT_DSL_STORE_CONFIG, topologyOverrides)) {
-            storeType = getString(DEFAULT_DSL_STORE_CONFIG);
+        // prefer dsl.store.suppliers.class to default.dsl.store and prefer topology overrides over
+        // the global streams configuration
+        if (isTopologyOverride(DSL_STORE_SUPPLIERS_CLASS_CONFIG, topologyOverrides)) {
+            storeSuppliers = (DslStoreSuppliers) Utils.newInstance(getClass(DSL_STORE_SUPPLIERS_CLASS_CONFIG));
+            log.info("Topology {} is overriding {} to {}", topologyName, DSL_STORE_SUPPLIERS_CLASS_CONFIG, storeSuppliers);
+        } else if (isTopologyOverride(DEFAULT_DSL_STORE_CONFIG, topologyOverrides)) {
+            final String storeType = getString(DEFAULT_DSL_STORE_CONFIG);
+            storeSuppliers = storeType.equals(IN_MEMORY) ? Materialized.StoreType.IN_MEMORY : Materialized.StoreType.ROCKS_DB;
             log.info("Topology {} is overriding {} to {}", topologyName, DEFAULT_DSL_STORE_CONFIG, storeType);
         } else {
-            storeType = globalAppConfigs.getString(DEFAULT_DSL_STORE_CONFIG);
+            storeSuppliers = (DslStoreSuppliers) Utils.newInstance(globalAppConfigs.getClass(DSL_STORE_SUPPLIERS_CLASS_CONFIG));
```

Comment:
I feel like we're missing a branch here -- if someone is using the old `default.dsl.store`  but only sets it in their global app configs (ie via `new KafkaStreams(...)`) and sets neither of these configs in their topology configs (ie `new StreamsBuilder(...)`), then we should still respect the old config and fall back on that in the global app configs